### PR TITLE
Ports Bay's updated evacuation controller and bluespace jump.

### PR DESCRIPTION
**Pull request by NonQueueingMatt**
Why are you looking at me like that? This is definitely not for the NBT.

### CODE REVIEW

**Comment by Tennessee116** [comment]
I'd like to suggest an optional follow-up implementation of Baystation12/Baystation12#26024 to go with the bluespace-jumping swagger.

**Comment by NonQueueingMatt** [comment]
> 
> 
> I'd like to suggest an optional follow-up implementation of [Baystation12/Baystation12#26024](https://github.com/Baystation12/Baystation12/pull/26024) to go with the bluespace-jumping swagger.

Definitely planned. Out of scope for this PR however.

**Comment by NonQueueingMatt** [comment]
!review

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+#define EVAC_PREPPING   1
+#define EVAC_LAUNCHING  2
+#define EVAC_IN_TRANSIT 3
+#define EVAC_COOLDOWN   4
+#define EVAC_COMPLETE   5
+
+var/datum/evacuation_controller/evacuation_controller
+
+/datum/evacuation_controller
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:11]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+/datum/evacuation_controller/proc/get_cooldown_message()
+	return "An evacuation cannot be called at this time. Please wait another [round((evac_cooldown_time-world.time)/600)] minute\s before trying again."
+
+/datum/evacuation_controller/proc/add_can_call_predicate(var/datum/evacuation_predicate/esp)
+	if(esp in evacuation_predicates)
+		CRASH("[esp] has already been added as an evacuation predicate")
+	evacuation_predicates += esp
+
+/datum/evacuation_controller/proc/call_evacuation(var/mob/user, var/_emergency_evac, var/forced, var/skip_announce, var/autotransfer)
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:61]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	if(state != EVAC_IDLE)
+		return 0
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:63]:
```suggestion
		return FALSE
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+	if(!can_evacuate(user, forced))
+		return 0
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:66]:
```suggestion
		return FALSE
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+		for(var/area/A in world)
+			if(istype(A, /area/hallway))
+				A.readyalert()
+		if(!skip_announce)
+			priority_announcement.Announce(replacetext(replacetext(current_map.emergency_shuttle_called_message, "%dock%", "[current_map.dock_name]"),  "%ETA%", "[round(get_eta()/60)] minute\s"), new_sound = 'sound/AI/emergencyshuttlecalled.ogg')
+	else
+		if(!skip_announce)
+			priority_announcement.Announce(replacetext(replacetext(current_map.shuttle_called_message, "%dock%", "[current_map.dock_name]"),  "%ETA%", "[round(get_eta()/60)] minute\s"), new_sound = 'sound/AI/shuttlecalled.ogg')
+
+	return 1
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:103]:
```suggestion
	return TRUE
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+/datum/evacuation_controller/proc/cancel_evacuation()
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:106]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+	if(!can_cancel())
+		return 0
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:108]:
```suggestion
		return FALSE
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	if(emergency_evacuation)
+		evac_recalled.Announce(current_map.emergency_shuttle_recall_message)
+		for(var/area/A in world)
+			if(istype(A, /area/hallway))
+				A.readyreset()
+		emergency_evacuation = 0
+	else
+		priority_announcement.Announce(current_map.shuttle_recall_message)
+
+	return 1
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:129]:
```suggestion
	return TRUE
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	state = EVAC_LAUNCHING
+
+	var/estimated_time = round(get_eta()/60,1)
+	if (emergency_evacuation)
+		evac_waiting.Announce(replacetext(current_map.emergency_shuttle_docked_message, "%ETA%", "[estimated_time] minute\s"), new_sound = sound('sound/AI/emergencyshuttledock.ogg', volume = 35))
+	else
+		priority_announcement.Announce(replacetext(replacetext(current_map.shuttle_docked_message, "%dock%", "[current_map.dock_name]"), "%ETA%", "[estimated_time] minute\s"), new_sound = sound('sound/AI/shuttledock.ogg'))
+
+/datum/evacuation_controller/proc/launch_evacuation()
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:141]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+		return
+
+	state = EVAC_IN_TRANSIT
+
+	if (emergency_evacuation)
+		priority_announcement.Announce(replacetext(replacetext(current_map.emergency_shuttle_leaving_dock, "%dock%", "[current_map.dock_name]"),  "%ETA%", "[round(get_eta()/60,1)] minute\s"))
+	else
+		priority_announcement.Announce(replacetext(replacetext(current_map.shuttle_leaving_dock, "%dock%", "[current_map.dock_name]"),  "%ETA%", "[round(get_eta()/60,1)] minute\s"))
+
+	return 1
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:152]:
```suggestion
	return TRUE
```

File: code/controllers/subsystems/evacuation/evacuation.dm [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+/datum/evacuation_controller/proc/finish_evacuation()
+	state = EVAC_COMPLETE
+
+/datum/evacuation_controller/process()
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation.dm:158]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation_eta.dm [hunk 12]
```diff
@@ -0,0 +1,22 @@
+/datum/evacuation_controller/proc/get_status_panel_eta()
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation_eta.dm:2]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation_helpers.dm [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	return state == EVAC_COMPLETE
+
+/datum/evacuation_controller/proc/is_on_cooldown()
+	return state == EVAC_COOLDOWN
+
+/datum/evacuation_controller/proc/is_evacuating()
+	return state != EVAC_IDLE
+
+/datum/evacuation_controller/proc/can_evacuate(var/mob/user, var/forced)
+
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation_helpers.dm:29]:
```suggestion
```

File: code/controllers/subsystems/evacuation/evacuation_helpers.dm [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+	if(!isnull(evac_called_at))
+		return 0
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation_helpers.dm:31]:
```suggestion
		return FALSE
```

File: code/controllers/subsystems/evacuation/evacuation_helpers.dm [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+	if (!universe.OnShuttleCall(null))
+		return 0
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation_helpers.dm:34]:
```suggestion
		return FALSE
```

File: code/controllers/subsystems/evacuation/evacuation_helpers.dm [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+	if(!forced)
+		for(var/predicate in evacuation_predicates)
+			var/datum/evacuation_predicate/esp = predicate
+			if(!esp.is_valid())
+				evacuation_predicates -= esp
+				qdel(esp)
+			else
+				if(!esp.can_call(user))
+					return 0
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation_helpers.dm:44]:
```suggestion
					return FALSE
```

File: code/controllers/subsystems/evacuation/evacuation_helpers.dm [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+	return 1
```
Comment (Geevies) [inline code/controllers/subsystems/evacuation/evacuation_helpers.dm:45]:
```suggestion
	return TRUE
```

**Review by Geevies — APPROVED** [review]
can you do the rest manually im dying here

**Review by Alberyk — APPROVED** [review]

**Pull request merged**